azure/scope/managedcontrolplane.go

```diff
@@ -102,13 +105,33 @@ type ManagedControlPlaneScope struct {
 	Client         client.Client
 	patchHelper    *patch.Helper
 	kubeConfigData []byte
+	cache          *ManagedControlPlaneCache
 
 	AzureClients
 	Cluster             *clusterv1.Cluster
 	ControlPlane        *infrav1exp.AzureManagedControlPlane
 	ManagedMachinePools []ManagedMachinePool
 }
 
+// ManagedControlPlaneCache stores common ManagedControlPlane information so we don't have to hit the API multiple times within the same reconcile loop.
+type ManagedControlPlaneCache struct {
+	isVnetManaged *bool
+}
+
+// InitManagedControlPlaneCache sets cached information about the ManagedControlPlane to be used in the scope.
+func (s *ManagedControlPlaneScope) InitManagedControlPlaneCache(ctx context.Context) error {
```

Comment:
I don't know if this init function is necessary given that we're fetching the info in IsVnetManaged() if cache is unset (seems redundant), we could probably get away with either:
a) having the init function set the cache at the beginning of the reconcile and then always return the cache value directly in IsVnetManaged() since cache would always be set, or
b) the first time IsVnetManaged() is called, it sets the value in the cache Just In Time and the rest of the time it gets the existing value from the cache

Reply:
I don't have a strong opinion in the abstract. In my view I'd take the practical approach: are there a non-trivial number of reconciliation contexts that won't actually need these cached values? If so, then definitely we should not init first.

I do have an opinion that we should do caching consistently across service scopes. In which case maybe we punt this as well to a separate workstream (not just vnet)? In my manual tests the impact of these vnet GET calls does not materially decrement the allowed requests limits at a speed necessary to contribute to subscription throttling, so I think we can reasonably land this improvement without necessarily needing to couple the caching optimization.

Reply:
> are there a non-trivial number of reconciliation contexts that won't actually need these cached values? If so, then definitely we should not init first.

the answer is no, every single reconcile loop will need at least call isVnetManaged at least once (both reconcile and delete) 

> I do have an opinion that we should do caching consistently across service scopes

agreed, do you have other examples of API GET calls that we need to do across multiple services that you're thinking of? We do have some redundant calls to isVnetManaged (especially one per subnet, which could grow as the number of subnets grows for large clusters with many node pools) so I do think caching is a worthy optimization this time around.

Reply:
I'm happy to scope the caching to this change (and this will be a good precedent going forward for net new changes that result in API calls).

The only reason I did the init is a basic copy/paste of the machine cache reference:

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/scope/machine.go#L102

I think doing these on-demand is the more intuitive approach (simpler as well), so I'll just modify this to do that.